ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1244 @@
+        }
+
+        return 1;
+    }
+
+    if (qc->aon_write_in_progress) {
+        /*
+         * AON write is in progress but we have not written everything yet. We
+         * may have managed to send zero bytes, or some number of bytes less
+         * than the total remaining which need to be appended during this
+         * AON operation.
+         */
+        qc->aon_buf_pos += actual_written;
+        assert(qc->aon_buf_pos < qc->aon_buf_len);
+        return QUIC_RAISE_NORMAL_ERROR(qc, SSL_ERROR_WANT_WRITE);
+    }
+
     /*
-     * TODO(QUIC): This is needed so the SSL_CTX_set_cipher_list("DEFAULT");
-     * produces at least one valid TLS-1.2 cipher.
-     * Later we should allow that there are none with QUIC protocol as
-     * SSL_CTX_set_cipher_list should still allow setting a SECLEVEL.
+     * Not in an existing AON operation but partial write is not enabled, so we
+     * need to begin a new AON operation. However we needn't bother if we didn't
+     * actually append anything.
      */
-    static const SSL_CIPHER ciph = {
-        1,
-        TLS1_TXT_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-        TLS1_RFC_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-        TLS1_CK_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-        SSL_kECDHE,
-        SSL_aRSA,
-        SSL_AES256GCM,
-        SSL_AEAD,
-        TLS1_2_VERSION, TLS1_2_VERSION,
-        DTLS1_2_VERSION, DTLS1_2_VERSION,
-        SSL_HIGH | SSL_FIPS,
-        SSL_HANDSHAKE_MAC_SHA384 | TLS1_PRF_SHA384,
-        256,
-        256
-    };
+    if (actual_written > 0)
+        aon_write_begin(qc, buf, len, actual_written);
 
-    return &ciph;
+    /*
+     * AON - We do not publicly admit to having appended anything until AON
+     * completes.
+     */
+    *written = 0;
+    return QUIC_RAISE_NORMAL_ERROR(qc, SSL_ERROR_WANT_WRITE);
 }
 
-int ossl_quic_renegotiate_check(SSL *ssl, int initok)
+static int quic_write_nonblocking_epw(QUIC_CONNECTION *qc, const void *buf, size_t len,
+                                      size_t *written)
 {
+    /* Simple best effort operation. */
+    if (!ossl_quic_sstream_append(qc->stream0->sstream, buf, len, written)) {
+        /* Stream already finished or allocation error. */
+        *written = 0;
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+    }
+
+    quic_post_write(qc, *written > 0, 1);
     return 1;
 }
 
-QUIC_CONNECTION *ossl_quic_conn_from_ssl(SSL *ssl)
+int ossl_quic_write(SSL *s, const void *buf, size_t len, size_t *written)
 {
-    return QUIC_CONNECTION_FROM_SSL(ssl);
+    QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_SSL(s);
+    int partial_write = ((qc->ssl_mode & SSL_MODE_ENABLE_PARTIAL_WRITE) != 0);
+
+    *written = 0;
+
+    if (!expect_quic_conn(qc))
+        return 0;
+
+    /* If we haven't started the handshake, do so automatically. */
+    if (!qc->started && !ossl_quic_do_handshake(qc))
+        return 0;
+
+    if (qc->ch != NULL && ossl_quic_channel_is_term_any(qc->ch))
```

Comment:
Is it ever possible for `qc->ch == NULL` here? Shouldn't that have been filled in by `ossl_quic_do_handshake()` above?